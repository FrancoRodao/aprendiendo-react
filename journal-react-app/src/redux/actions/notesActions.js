import { db } from "../../firebase/firebase.config"
import { imageUploadToCloudinary } from "../../helpers/imageUpload"
import { loadNotes } from "../../helpers/loadNotes"
import { types } from "../types/types"

export const startAddNewNote = () => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        const newNote = {
            title: 'title',
            body: 'body',
            date: new Date().getTime()
        }

        const noteRef = await db.collection(`${userId}/journal/notes`).add(newNote)
        dispatch(addNewNote(noteRef.id, newNote))
        dispatch(activeNoteAction(noteRef.id, newNote))
    }
}

export const addNewNote = (noteId, note) => ({
    type: types.notesAdd,
    payload: {
        id: noteId,
        note
    }
})

export const activeNoteAction = (id, note) => ({
    type: types.notesSetActive,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (userId) => {
    return async (dispatch) => {
        const notes = await loadNotes(userId)
        dispatch(setNotes(notes))
    }
}

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: notes
})


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth
        const noteToFireStore = {...note}
        delete noteToFireStore.id

        await db.collection(`${uid}/journal/notes`).doc(`${note.id}`).update(noteToFireStore)
        dispatch(refreshNote(note))
    }
}

export const refreshNote = (note) => ({
    type: types.notesUpdate,
    payload: {
        note
    }
})

export const startUploadPicture = (file) => {
    return async (dispatch, getState) => {
        const activeNote = getState().notes.active
        const imageUrl = await imageUploadToCloudinary(file)
        activeNote.imgUrl = imageUrl
        
        dispatch(startSaveNote(activeNote))
    }
}


export const startDeleteNote = (note) => {
    return async (dispatch, getState) => {
        const userId = getState().auth.uid

        await db.doc(`${userId}/journal/notes/${note.id}`).delete()
        dispatch(deleteNote(note))
    }
}

export const deleteNote = (note) => ({
    type: types.notesDelete,
    payload: {
        note
    }
})