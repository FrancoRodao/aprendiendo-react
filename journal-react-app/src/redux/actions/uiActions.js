import { types } from "../types/types"

export const startLoadingAction = () => (
    {
        type: types.uiStartLoading,
        payload: {
            loading: true
        }
    }
)


export const finishLoadingAction = () => (
    {
        type: types.uiFinishLoading,
        payload: {
            loading: false
        }
    }
)

export const errorAction = (state, message) => (
    {
        type: types.uiError,
        payload: {
            state,
            lastMsg: message
        }
    }
)

export const clearErrorsAction = () =>  (
    {
        type: types.uiError,
        payload: {
            state: false,
            lastMsg: ""
        }
    }
)