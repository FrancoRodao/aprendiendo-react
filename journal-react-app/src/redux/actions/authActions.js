import { types } from "../types/types"
import { firebase, googleAuthProvider } from "../../firebase/firebase.config";
import { clearErrorsAction, errorAction, finishLoadingAction, startLoadingAction } from "./uiActions";

export const startLogin = (email, password) => {
    return (dispatch) => {
        dispatch(startLoadingAction())

        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(e => {
                if (e.code === "auth/user-not-found" || "auth/wrong-password") {
                    dispatch(errorAction(true, "Wrong credentials"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const startGoogleLogin = () => {
    return dispatch => {
        dispatch(startLoadingAction())

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                dispatch(loginAction(user.uid, user.displayName))
                dispatch(clearErrorsAction())
            })
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const loginAction = (uid, displayname) => (
    {
        type: types.authLogin,
        payload: {
            uid,
            displayname
        }
    }
)

export const startRegisterUser = (name, email, password) => {
    return (dispatch, state) => {
        dispatch(startLoadingAction())
        return firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async ({ user }) => {
                await user.updateProfile({
                    displayName: name,
                    photoURL: ''
                })
                dispatch(registerUserAction(user.uid, user.displayName))
                dispatch( clearErrorsAction() )
            })
            .catch(e => {
                if (e.code === "auth/email-already-in-use") {
                    dispatch(errorAction(true, "Email already registered"))
                    return
                }
                dispatch(errorAction(true, "Unexpected error"))
            })
            .then(() => state())
            .finally(() => dispatch(finishLoadingAction()))
        
    }
}

export const registerUserAction = (uid, displayname) => {
    return {
        type: types.register,
        payload: {
            uid,
            displayname
        }
    }
}


export const startLogout = () => {
    return async (dispatch) => {
        dispatch(startLoadingAction())

        firebase.auth().signOut()
            .then(() => dispatch(logoutUserAction()))
            .catch(() => dispatch(errorAction(true, "Unexpected error")))
            .finally(() => dispatch(finishLoadingAction()))
    }
}

export const logoutUserAction = () => ({
    type: types.authLogout,
    payload: {}
})

