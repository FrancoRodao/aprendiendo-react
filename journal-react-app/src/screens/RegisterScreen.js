import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from "validator";
import { startRegisterUser } from '../redux/actions/authActions';
import { clearErrorsAction } from '../redux/actions/uiActions';

export default function RegisterScreen({ history }) {

    const [formErrors, setFormErrors] = useState({
        name: '',
        email: '',
        password: ''
    })

    const dispatch = useDispatch()
    const state = useSelector(state => state)
    const loading = state.ui.loading
    const {state: errorState, msgErr: errorUIMessage } = state.ui.errors
    
    const handleSubmit = (e) => {
        e.preventDefault()
        const form = {
            name: e.target[0].value,
            email: e.target[1].value,
            password: e.target[2].value,
            confirmPassword: e.target[3].value
        }

        const { name, email, password } = form

        if (isFormValid(form)) {
            dispatch( startRegisterUser(name, email, password) ).then( state => {
                if(state.ui.errors.state) return
                history.push("/")
            })
        }

    }

    const isFormValid = ({ name, email, password, confirmPassword }) => {

        let formErrors = {
            name: '',
            email: '',
            password: ''
        }

        if (name.length === 0) {
            formErrors.name = "Name is required"
        }

        if (!validator.isEmail(email)) {
            formErrors.email = "Email is not valid"
        }

        if (password !== confirmPassword || password.length < 5) {
            formErrors.password = "Password's should be at least 6 characters and match each other"
        }

        setFormErrors(formErrors)

        if (formErrors.name.length > 0 || formErrors.email.length > 0 || formErrors.password.length > 0) {
            return false
        } else {
            return true
        }

    }

    //clear errors when change page
    const handleLinkClick = () => dispatch( clearErrorsAction() )

    return (
        <>
            <h3 className={`auth__title ${errorState ? 'auth__title-errorUI' : ''}`}>
                Register
            </h3>

            {
                errorState &&
                <div className="auth__alert-error-ui">
                    {errorUIMessage}
                </div>
            }

            <form onSubmit={handleSubmit} autoComplete="off">
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    autoComplete="off"
                    className={`auth__input ${formErrors.name.length > 0 ? 'auth__input-error' : ''}`} />

                {
                    formErrors.name.length > 0 &&
                    <p className="auth__alert-error">
                        {formErrors.name}
                    </p>
                }

                <input
                    type="email"
                    placeholder="Email"
                    name="email"
                    autoComplete="on"
                    className={
                        `auth__input ${formErrors.email.length > 0 ?
                            'auth__input-error' : ''}`
                    }
                />

                {
                    formErrors.email.length > 0 &&
                    <p className="auth__alert-error">
                        {formErrors.email}
                    </p>
                }

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className={
                        `auth__input 
                        ${formErrors.password.length > 0 ?
                            'auth__input-error' : ''}`
                    } />

                {
                    formErrors.password.length > 0 &&
                    <p className="auth__alert-error">
                        {formErrors.password}
                    </p>
                }

                <input
                    type="password"
                    placeholder="Confirm password"
                    name="confirmPassword"
                    className={
                        `auth__input ${formErrors.password.length > 0 ?
                            'auth__input-error' : ''}`
                    }
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                >
                    {
                        loading ?
                            <i className="fas fa-spinner fa-spin"></i>
                            :
                            "Register"
                    }
                </button>

                <hr className="mb-5"></hr>

                <Link onClick={handleLinkClick} className="link" to="/login" >
                    Already registered?
                </Link>

            </form>
        </>
    )
}
