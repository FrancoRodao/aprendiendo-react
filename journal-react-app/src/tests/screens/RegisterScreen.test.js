import { mount } from "enzyme"
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import RegisterScreen from "../../screens/RegisterScreen";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import { startRegisterUser } from "../../redux/actions/authActions";
import { clearErrorsAction } from "../../redux/actions/uiActions";

jest.mock('../../redux/actions/authActions', () => ({
    __esModule: true,
    startRegisterUser: jest.fn(),
}))

jest.mock('../../redux/actions/uiActions', () => ({
    __esModule: true,
    clearErrorsAction: jest.fn()
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const storeInitialState = {
    auth: {},
    notes: {},
    ui: {
        loading: false,
        errors: {
            state: false,
            msgErr: ''
        }
    }
}
let store = mockStore(storeInitialState)
store.dispatch = () => Promise.resolve({
    ui:{
        errors: {
            state: false
        }
    }
})

describe('Register screen tests', () => {

    beforeEach(() => store = mockStore(storeInitialState))

    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen history={{
                    push: jest.fn()
                }} />
            </MemoryRouter>
        </Provider>
    )

    test('should match to snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('should dispatch start register user', () => {

        wrapper.find('.btn.btn-primary.btn-block').simulate("submit", {
            preventDefault: () => { },
            target: [
                {
                    value: 'new-user'
                },
                {
                    value: 'new-user@gmail.com'
                },
                {
                    value: 'new-user-password'
                },
                {
                    value: 'new-user-password'
                }
            ]
        })

        expect(startRegisterUser)
            .toHaveBeenCalledWith(
                'new-user', 'new-user@gmail.com', 'new-user-password'
            )
    })

    test('should show error', () => {
        const storeInitialState = {
            auth: {},
            notes: {},
            ui: {
                loading: false,
                errors: {
                    state: true,
                    msgErr: 'EMAIL ALREADY REGISTER OR SOME STUPID ERROR XD'
                }
            }
        }
        const store = mockStore(storeInitialState)

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        )

        const errorText = wrapper.find('.auth__alert-error-ui').text()
        expect(errorText).toBe(storeInitialState.ui.errors.msgErr)
    })
    

    test('should clear errors when change page', () => {
        wrapper.find("Link").simulate("click")
        expect(clearErrorsAction).toBeCalledTimes(1)
    })
    

})
