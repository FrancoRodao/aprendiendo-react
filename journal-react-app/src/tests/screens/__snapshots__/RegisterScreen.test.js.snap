// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register screen tests should match to snapshot 1`] = `
Array [
  <h3
    className="auth__title "
  >
    Register
  </h3>,
  <form
    autoComplete="off"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input "
      name="name"
      placeholder="Name"
      type="text"
    />
    <input
      autoComplete="on"
      className="auth__input "
      name="email"
      placeholder="Email"
      type="email"
    />
    <input
      className="auth__input 
                        "
      name="password"
      placeholder="Password"
      type="password"
    />
    <input
      className="auth__input "
      name="confirmPassword"
      placeholder="Confirm password"
      type="password"
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <hr
      className="mb-5"
    />
    <a
      className="link"
      href="/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
